File: hard/76.minimum-window-substring.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map.Entry;

/*
 * @lc app=leetcode id=76 lang=java
 *
 * [76] Minimum Window Substring
 */

// @lc code=start
class Solution {
    private HashMap<Character, Integer> targetMap = new HashMap<>();
    private HashMap<Character, Integer> windowMap = new HashMap<>();

    public String minWindow(String s, String t) {
        char[] chars = s.toCharArray();
        int l = 0;
        int ansL = -1;
        int ansR = -1;

        char[] tChars = t.toCharArray();
        for (char c : tChars) {
            targetMap.put(c, targetMap.getOrDefault(c, 0) + 1);
        }
        
        int minLen = Integer.MAX_VALUE;
        for (int r=0; r<chars.length; r++) {
            if (targetMap.containsKey(chars[r])) {
                windowMap.put(chars[r], windowMap.getOrDefault(chars[r], 0) + 1);
            }
            while (targetExists() && l <= r) {
                // this window has all the chars in t string, now move left to shrink window size
                int windowLen = r - l + 1;
                if (windowLen < minLen) {
                    // update the result
                    minLen = windowLen;
                    ansR = r + 1;
                    ansL = l;
                }
                if (targetMap.containsKey(chars[l])) {
                    windowMap.put(chars[l], windowMap.getOrDefault(chars[l], 0) - 1);
                }

                l++;
            }
        }

        // ""cabwefgewcwaefgcf"\n"cae""
        return ansL == -1 ? "" : s.substring(ansL, ansR);
    }

    private boolean targetExists() {
        for (Map.Entry<Character, Integer> entry : targetMap.entrySet()) {
            if (windowMap.getOrDefault(entry.getKey(), 0) < entry.getValue()) {
                return false;
            }
        }

        return true;
    }

    private String bruteForce(String s, String t) {
        if (s.length() < t.length()) {
            return "";
        }

        String res = "";
        char[] chars = s.toCharArray();
        for (int i=0; i<chars.length; i++) {
            StringBuffer sb = new StringBuffer(t);
            for (int j=i; j<chars.length; j++) {
                if (sb.length() == 0) {
                    res = getMinString(res, s, i, j);
                    break;
                }
                int index = sb.toString().indexOf(chars[j]);
                if (index != -1) {
                    sb.deleteCharAt(index);
                }
            }
        }

        return res;
    }

    private String getMinString(String res, String s, int i, int j) {
        if (!res.isEmpty() && s.length() <= j - i + 1) {
            return res;
        } else {
            return s.substring(i, j+1);
        }
    }
}
// @lc code=end

File: hard/76.minimum-window-substring.py
#
# @lc app=leetcode id=76 lang=python3
#
# [76] Minimum Window Substring
#

# @lc code=start
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        def match():
            for i in range(len(chars_in_t)):
                if chars_in_t[i] > chars_in_window[i]:
                    return False
            return True

        if len(s) < len(t):
            return ""

        chars_in_t = [0] * 58
        chars_in_window = [0] * 58

        for c in t:
            chars_in_t[ord(c) - ord('A')] += 1

        min_len = len(s) + 1
        min_l = 0
        min_r = 0

        l = 0
        for r in range(len(s)):
            chars_in_window[ord(s[r]) - ord('A')] += 1
            while match():
                cur_len = r - l + 1
                if cur_len < min_len:
                    min_len = cur_len
                    min_l = l
                    min_r = r
                chars_in_window[ord(s[l]) - ord('A')] -= 1
                l += 1
        
        return "" if min_len == len(s) + 1 else s[min_l : min_r+1]

""" top voted
def minWindow(self, s, t):
    need, missing = collections.Counter(t), len(t)
    i = I = J = 0
    for j, c in enumerate(s, 1):
        missing -= need[c] > 0
        need[c] -= 1
        if not missing:
            while i < j and need[s[i]] < 0:
                need[s[i]] += 1
                i += 1
            if not J or j - i <= J - I:
                I, J = i, j
    return s[I:J]
"""
# @lc code=end

